Allow multi-line strings using `str` in translation macros

noahmoss opened the pull request:
This is a proposal for a potential way to solve #10770, which is an issue that I've seen come up a couple times now for me and others.

Basically, this change means that the `tru`, `trs`, `deferred-tru` and `deferred-trs` macros can take as a first argument _either_ a string, or a `str` form that has all strings as arguments. If the latter is used, the strings are concatenated (using `str`) before the `UserLocalizedString`/`SiteLocalizedString` is created. The `str` is basically just syntactic sugar; it could equally be a vector of strings that are implicitly joined or something similar, but this syntax seemed intuitive.

This allows us to break up long strings over multiple lines but have them translated as a single unit. It also fixes some bugs where `str` was being used to join multiple `deferred-tru` macros in code that was executed at compile-time, thus losing the advantage of a deferred translation lookup. This happened in a couple setting descriptions, so I've converted those to use the new syntax as examples (as well as a few others that were exceeding 120 chars).

noahmoss committed:
allow str form as the first argument to deferred-tru

noahmoss committed:
add tests and update doc strings

noahmoss committed:
migrate more setting descriptions

noahmoss commented:
@metamben Here's a potential alternative solution for some of the translation macro workarounds you're using in #22804.

metamben reviewed (approved):
If we prefer to have the text to be translated in a single chunk, then I think this solution is good. Also, my understanding is that after this change we will have the possibility to concatenate string literals where we had to write a single string literal, so this is an extension that should not break existing code.

camsaul reviewed (changes requested):
The one line restriction isn't arbitrary, it's because the tool we use to scrape our source files doesn't support multi-line forms.

https://github.com/metabase/metabase/blob/master/bin/i18n/update-translation-template#L53-L69

If you want to support multi-line forms you'll have to replace that tool with something more sophisticated. I would write some custom Clojure code that walks the source files without evaluating anything and then looks for the `tr*`/`deferred-tr*` forms and can handle `str` inside them. This would be nice because it would let us remove our dependency on `xgettext` for running the build scripts.

noahmoss commented:
@camsaul Ah, thanks for the context. I wasn't aware that's how the script worked...might try to tackle that at some point but probably not right away. Will close this PR in the meantime.

dpsutton commented:
@noahmoss in light of https://github.com/metabase/metabase/pull/23181, if that PR is merged can this PR be merged as-is or would I need to do anything?

noahmoss commented:
@dpsutton I think this PR should be good to go when that one is merged w/support for `(str ...)` forms.

noahmoss commented:
Or we could update this PR to add that support, if that would compartmentalize the changes better. I don't have a strong preference.

dpsutton commented:
So just merged the new pot creation script. I can finish this branch unless someone else wants to jump on. the fix is quite simple. In the file `bin/i18n/src/i18n/enumerate.clj` there's a function:

```clojure
(defn- form->string-for-translation
  "Function that turns a form into the translation string. At the moment
  it is just the second arg of the form. Afterwards it will need to
  concat string literals in a `(str \"foo\" \"bar\")` situation. "
  [form]
  (second form))
```

Form at this point will be a form like `(trs "Invalid foo: {0}" foo)` and right now it just takes `second`, in this case `"Invalid foo: {0}"`. We can update this to take second if second is a string, otherwise anticipate a `(str "string literal 1" "string literal 2")` and grab the literal forms and combine them.

dpsutton committed:
Handle multiline translation sources

```clojure
enumerate=> x
(deferred-tru
 (str
  "Whether an introductory modal should be shown after the next database connection is added. "
  "Defaults to false if any non-default database has already finished syncing for this instance."))
enumerate=> (form->string-for-translation x)
"Whether an introductory modal should be shown after the next database connection is added. Defaults to false if any non-default database has already finished syncing for this instance."
```

We get the form from `(g/grasp <file> ::translate)` which returns the
`x`. And then simply pick through it for either a string literal or a
call to `(str <literal>+)`

dpsutton commented:
@camsaul can you mark your requested changes as completed?

```
;; the pot file
#: metabase/public_settings.clj:521
msgid ""
"Whether an introductory modal should be shown after the next database "
"connection is added. Defaults to false if any non-default database has "
"already finished syncing for this instance."
msgstr ""
```

```clojure
;; source
(defsetting show-database-syncing-modal
  (deferred-tru
    (str "Whether an introductory modal should be shown after the next database connection is added. "
         "Defaults to false if any non-default database has already finished syncing for this instance."))
  :visibility :admin
  :type       :boolean
  :getter     (fn []
                (let [v (setting/get-value-of-type :boolean :show-database-syncing-modal)]
                  (if (nil? v)
                    (not (db/exists? 'Database :is_sample false, :initial_sync_status "complete"))
                    ;; frontend should set this value to `true` after the modal has been shown once
                    v))))
```
